shared/deps/serial/serial_host_send_receive.py

```diff
@@ -156,9 +156,23 @@ def send(connect, filename, p_size=1024):
         print("Sent size = %s" % send_size)
         file_no.close()
     md5_sum = md5_value.hexdigest()
+    wait_receive_ack(connect)
     return md5_sum
 
 
+def wait_receive_ack(connect):
+    exp_str = b'ALLRECEIVED'
+    exp_size = len(exp_str)
+    r_str = b''
+    r_size = 0
+    while r_size < exp_size:
+        txt = connect.recv(exp_size)
+        r_size += len(txt)
+        r_str += txt
+    if struct.unpack("11s", txt)[0] != exp_str:
+        raise ShakeHandError("Not get Received ACK from guest.")
```

Comment:
I think we need that works on dual side, I mean both `host -> guest` and `guest -> host`, right?

Reply:
@luckyh I have thought about that, but finally not add `guest->host` side because 1. Host has receive buffer so lower risk for data loss; 2. I was worried about the process would hang at any while loop if meet exception(like previous hang at guest receive loop), one more while loop more risk for that. So I prefer not adding more while loop if not very necessary. But still I am open to it, let me know your thought please, thanks.

Reply:
> @luckyh I have thought about that, but finally not add `guest->host` side because 1. Host has receive buffer so lower risk for data loss;

So that means there is the probability to get an automation failure, right? then @yanan-fu, what do you think about that?

> 2. I was worried about the process would hang at any while loop if meet exception(like previous hang at guest receive loop), one more while loop more risk for that. So I prefer not adding more while loop if not very necessary. But still I am open to it, let me know your thought please, thanks.

Well, it doesn't sound like a problem, actually you will either get md5 mismatch or transmission timeout failures when the exception happened. And, to prevent tests getting stuck we can simply give a reasonable `timeout` value to `session.cmd`/`process.run`.

Reply:
@luckyh I have tried let guest also wait for host ack after receive data, with current version code. But it introduce new issue: Host script will exit after send its 'receive ack', so sometimes guest will loss this 'ack',  and go into infinite waiting loop, finally lead guest session timeout and fail the case. That's not 100% reproducible but have very high chance to hit(2/3 in my test today). I think for current implementation it's not a good idea to let guest wait more. And I think the 'probability' for host to loss data is very very low, we can ignore it. What do you think?

Reply:
@luckyh  @vivianQizhu  After checking, I agree with xuhan more about check `guest -> host`,  same with `guest -> host`. what's more, I can't understand well why the 'probability' for host to loss data is low, but not guest.
And for host script exit after send 'receive ack', it is should be a bug which need be fix rather than bypass it.

Reply:
@yanan-fu I believe that's because host use socket so it has receive buffer, while guest use os.read() so no receive buffer. Now let guest wait receive ack from host increases the chance guest loss data(the ack message), It failed a lot in my test.
1. We only hit data loss from guest side currently, by adding the receive ack for `host-> guest` it does not fail anymore.
2. Never hit data loss from host so far(test more than 100 times).
3. Add receive ack for `guest->host`, case will fail some times cause host script exit after send ack message but before guest receive it.
In one word, your request will not fix any issue but only highly increase the fail ratio of this case.

Reply:
And the 'probability' for host to loss data, currently it only exists in theory, never really hit it, so I say it is very rare. While the guest loss data is hit a lot of times during my recent test.

Reply:
> @yanan-fu I believe that's because host use socket so it has receive buffer, while guest use os.read() so no receive buffer. Now let guest wait receive ack from host increases the chance guest loss data(the ack message), It failed a lot in my test.
> 
>     1. We only hit data loss from guest side currently, by adding the receive ack for `host-> guest` it does not fail anymore.
> 
>     2. Never hit data loss from host so far(test more than 100 times).
> 
>     3. Add receive ack for `guest->host`, case will fail some times cause host script exit after send ack message but before guest receive it.

My point is, this is a issue need be fixed. we can not miss the checking step, just because the change can cause new issue. Thanks!

>        In one word, your request will not fix any issue but only highly increase the fail ratio of this case.

Reply:
@yanan-fu No, don't be confused, there is no such steps, add receive ack is to fix issue, it's not a checking step. The only thing we need to check is that sender and receiver has same md5 value after file transfter, which has to be accomplished. If there is no guest data loss issue(that's introduced by previous code, cause it uses os.read, and it has to use it), we don't need the receive ack at all. 
Actually my previous solution is to set small send buffer to make host sender slower, to match the rhythm of guest receiver, will can reduce the data loss, just in theory it still has the risk.

Reply:
@yanan-fu To be more clear, the issue we need to fix is: guest will loss data if host scrip exit before guest finish all receive action, cause guest has no receive buffer. So the solution is let host wait for guest tells that it finish all data receive action: Add a receive ack after guest receive all data.
Host has no such issue, and because the guest no buffer issue, let guest wait host's ack will cause the same problem: guest might loss the ack message if host exit first.
We can't make host exit first before guest receive all data, that's why we can't let guest wait for host's ack. Hope I explained it better this way.

Reply:
@yanan-fu Based on the above situation, if you have better solution, that will be appreciated. Thanks.

Reply:
@vivianQizhu my opinion remains the same,  since i still think `guest -> host` and `host -> guest` are same in theory, to avoid the risk.  However, you can insist yours if you think the risk is acceptable from feature testing perspective. Thanks

Reply:
@yanan-fu I have to say even in theory they are not same, cause their read/write method are totally not same.